### PR TITLE
LIME-1247 - Add SSM Parameter Cache Lambda

### PR DESCRIPTION
**Pull request by smsgds**
### What changed

	- Disabled integration tests that tested errors for invalid QuestionURL/OTG Url as the correct values will be cached in previous journeys and cannot be overriden until the cache expires in the lambda.

	- Add SSMParameterRequestFailureAlarm triggering when SSMParametersFunction fails for any reason.

	- Add SSMParametersFunction to FetchQuestionsStateMachine, PostAnswerStateMachine, IssueCredentialStateMachine
	- - Add GetParameters permissions to SSMParametersFunction
	- - Accounts for Prefixes Stackname, ParameterPrefix and CommonStackName
	- - Add StrategyUtil for test data strategy (note this PR does not implement test data strategy)
	- - Add Unit tests for SSMParametersFunction + StrategyUtil

	- FetchQuestionsStateMachine / PostAnswerStateMachine
	- - Align DefinitionSubstitutions arrangement.
	- - Remove known unused Definition Substitutions
	- - Remove Parameter permissions
	- - Switch to using SSMParametersFunction

	- FetchQuestionsHandler
	- - RetrieveQuestinoService - changed url to questionsUrl to make it easier to catch parameter value missmatch

	- SubmitAnswerHandler
	- - SubmitAnswerService - changed url to answersUrl to make it easier to catch parameter value missmatch

	- PostAnswerStateMachine
	- - Remove Parameter permissions
	- - Remove known unused Definition Substitutions
	- - Switch to using SSMParametersFunction

	- IssueCredentialStateMachine
	- - Remove Parameter permissions
	- - Remove known unused Definition Substitutions
	- - Switch to using SSMParametersFunction

	- IssueCredentialLambda
	- - Changed verifiableCredentialssuer to issuer as name is programatically set from the parameter name.
	- - Changed kmsSigningKeyId to verifiableCredentialKmsSigningKeyId as name is programatically set from the parameter name.

	- re-enable coverageReporters

### Why did it change

To Optimize SSM parameter reads aiming to avoid hitting SSM rate limits

### Issue tracking

- [LIME-1247](https://govukverify.atlassian.net/browse/LIME-1247)

### Other considerations

[LIME-1247]: https://govukverify.atlassian.net/browse/LIME-1247?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: infrastructure/template.yaml [hunk 1]
```diff
@@ -1198,6 +1128,61 @@ Resources:
...
+      LogGroupName: !Sub "/aws/lambda/${SSMParametersFunction}"
+      RetentionInDays: 30
+
+  SSMParametersFunctionLogGroupSubscriptionFilterCsls:
+    Type: AWS::Logs::SubscriptionFilter
+    Condition: LogSendingEnabled
+    Properties:
+      DestinationArn: "arn:aws:logs:eu-west-2:885513274347:destination:csls_cw_logs_destination_prodpython-2"
+      FilterPattern: ""
+      LogGroupName: !Ref SSMParametersFunctionLogGroup
```
Comment (gds-jbentham) [inline infrastructure/template.yaml:1184]: Will this need onboarded to CSLS or will it be caught by our existing filters?

Comment (smsgds) [reply infrastructure/template.yaml:1184]: no change needed the log group path matches the other lambdas `/aws/lambda/`

File: lambdas/ssm-parameters/src/ssm-parameters-handler.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  private metricsProbe: MetricsProbe;
+
+  private randomMaxAgeInSeconds: number;
+
+  constructor(metricsProbe: MetricsProbe) {
+    this.metricsProbe = metricsProbe;
+
+    // Must calulate in seconds
+    const randomValue = Math.random(); // NOSONAR
+    this.randomMaxAgeInSeconds = Math.floor(
```
Comment (ElliotMurphyGDS) [inline lambdas/ssm-parameters/src/ssm-parameters-handler.ts:36]: Why are we making max age a random value here? Surely we want it to be a controlled length of time/value?

Comment (smsgds) [reply lambdas/ssm-parameters/src/ssm-parameters-handler.ts:36]: When scaling, there will be multiple lambdas in play. We do not want them all expiring the parameter cache at the same time. In the scenario where we have high traffic and lots of scaled ssm lambdas when these expired we would hit the SSM Rate limits as the caches expire at the same time - the random offset avoid thats.

This is a direct implementation of the same behaviour of the Java SSM caching - 
https://github.com/govuk-one-login/ipv-cri-lib/blob/main/src/main/java/uk/gov/di/ipv/cri/common/library/util/ClientProviderFactory.java#L152

File: infrastructure/template.yaml [hunk 3]
```diff
@@ -1374,6 +1359,30 @@ Resources:
...
+        - !Ref AlarmTopicHmrcKbv
+      InsufficientDataActions: []
+      MetricName: CompletionStatus
+      Namespace: !Ref CriIdentifier
+      Statistic: Minimum
+      Dimensions:
+        - Name: service
+          Value: !Sub "${CriIdentifier}-ssm-parameters"
+      Period: 60
+      EvaluationPeriods: 1
```
[the remark this answers is not in the document]
Comment (smsgds) [reply infrastructure/template.yaml]: Have changed this to an average below 0.95 (5%) over three minutes. To account for any temporary blip reading parameters.

**Review by gds-jbentham — APPROVED** [review]

**Pull request merged**